File: banchus-app/src/main/java/pe/nanamochi/banchus/service/StorageService.java
package pe.nanamochi.banchus.service; // O pe.nanamochi.banchus.service.infra según prefieras

import java.io.IOException;
import java.nio.file.Path;
import java.util.List;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import pe.nanamochi.banchus.domain.storage.FileStorageProvider;
import pe.nanamochi.banchus.util.Security;

@Slf4j
@Service
@RequiredArgsConstructor
public class StorageService {
  private final FileStorageProvider provider;

  private static final String AVATARS = "avatars_files";
  private static final String BEATMAPS = "osu_beatmap_files";
  private static final String REPLAYS = "replays_files";
  private static final String SCREENSHOTS = "screenshots_files";

  public static final List<String> ALL_BUCKETS = List.of(AVATARS, BEATMAPS, REPLAYS, SCREENSHOTS);

  public void initStorage() throws IOException {
    provider.initialize(ALL_BUCKETS);

    if (!provider.exists(AVATARS, "default.png")) {
      try (var is = getClass().getResourceAsStream("/images/default.png")) {
        if (is != null) {
          provider.write(AVATARS, "default.png", is.readAllBytes());
          log.info("Default avatar initialized.");
        }
      } catch (IOException e) {
        log.error("Failed to setup default avatar", e);
      }
    }
  }

  public byte[] getAvatar(String userId) {
    return provider
        .read(AVATARS, userId + ".png")
        .or(() -> provider.read(AVATARS, "default.png"))
        .orElse(new byte[0]);
  }

  public void uploadAvatar(String userId, byte[] content) {
    provider.write(AVATARS, userId + ".png", content);
  }

  public Optional<byte[]> getBeatmap(int beatmapId) {
    return provider.read(BEATMAPS, beatmapId + ".osu");
  }

  public void uploadBeatmap(int beatmapId, byte[] content) {
    provider.write(BEATMAPS, beatmapId + ".osu", content);
  }

  public boolean beatmapExists(int beatmapId) {
    return provider.exists(BEATMAPS, beatmapId + ".osu");
  }

  public Path getBeatmapPath(int beatmapId) {
    return provider.getFileAsPath(BEATMAPS, beatmapId + ".osu");
  }

  public Optional<byte[]> getReplay(long scoreId) {
    return provider.read(REPLAYS, scoreId + ".osr");
  }

  public void saveReplay(long scoreId, byte[] content) {
    provider.write(REPLAYS, scoreId + ".osr", content);
  }

  public Optional<byte[]> getScreenshot(String screenshotId) {
    return provider.read(SCREENSHOTS, screenshotId + ".png");
  }

  public String saveScreenshot(byte[] content) {
    String id = Security.generateToken(6);
    provider.write(SCREENSHOTS, id + ".png", content);
    return id;
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/service/StorageService.kt
package pe.nanamochi.banchus.service

import com.github.michaelbull.result.Result
import com.github.michaelbull.result.orElse
import com.github.michaelbull.result.toResultOr
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import pe.nanamochi.banchus.domain.errors.FileNotFound
import pe.nanamochi.banchus.domain.errors.StorageError
import pe.nanamochi.banchus.domain.storage.FileStorageProvider
import pe.nanamochi.banchus.util.Security
import pe.nanamochi.banchus.util.runStorageCatching

@Service
class StorageService(private val provider: FileStorageProvider) {
    private val log = LoggerFactory.getLogger(javaClass)

    companion object {
        const val AVATARS = "avatars_files"
        const val BEATMAPS = "osu_beatmap_files"
        const val REPLAYS = "replays_files"
        const val SCREENSHOTS = "screenshots_files"

        private val ALL_BUCKETS = listOf(AVATARS, BEATMAPS, REPLAYS, SCREENSHOTS)
    }

    fun initStorage() {
        provider.initialize(ALL_BUCKETS)

        if (!provider.exists(AVATARS, "default.png")) {
            javaClass
                .getResourceAsStream("/images/default.png")
                ?.use { it.readAllBytes() }
                ?.let { bytes ->
                    provider.write(AVATARS, "default.png", bytes)
                    log.info("Default avatar initialized successfully.")
                } ?: log.warn("System resource /images/default.png not found!")
        }
    }

    fun getAvatar(userId: String): Result<ByteArray, FileNotFound> =
        provider
            .read(AVATARS, userId.asPng())
            .toResultOr { FileNotFound }
            .orElse { provider.read(AVATARS, "default.png").toResultOr { FileNotFound } }

    fun saveAvatar(userId: String, content: ByteArray): Result<Unit, StorageError> =
        runStorageCatching {
            provider.write(AVATARS, userId.asPng(), content)
        }

    fun getBeatmap(beatmapId: Int): Result<ByteArray, FileNotFound> =
        provider.read(BEATMAPS, beatmapId.asOsu()).toResultOr { FileNotFound }

    fun saveBeatmap(beatmapId: Int, content: ByteArray): Result<Unit, StorageError> =
        runStorageCatching {
            provider.write(BEATMAPS, beatmapId.asOsu(), content)
        }

    fun beatmapExists(beatmapId: Int): Boolean = provider.exists(BEATMAPS, beatmapId.asOsu())

    fun getReplay(scoreId: Long): Result<ByteArray, FileNotFound> =
        provider.read(REPLAYS, scoreId.asOsr()).toResultOr { FileNotFound }

    fun saveReplay(scoreId: Long, content: ByteArray): Result<Unit, StorageError> =
        runStorageCatching {
            provider.write(REPLAYS, scoreId.asOsr(), content)
        }

    fun getScreenshot(screenshotId: String): Result<ByteArray, FileNotFound> =
        provider.read(SCREENSHOTS, screenshotId.asPng()).toResultOr { FileNotFound }

    fun saveScreenshot(content: ByteArray): Result<String, StorageError> = runStorageCatching {
        Security.generateToken(6).let { id ->
            provider.write(SCREENSHOTS, id.asPng(), content)
            id
        }
    }

    private fun Any.asPng() = "$this.png"

    private fun Any.asOsu() = "$this.osu"

    private fun Any.asOsr() = "$this.osr"
}
